src/eos-prepare-usb-update.c

```diff
@@ -0,0 +1,665 @@
+{
+  g_clear_object (&pull_data->progress);
+  g_clear_object (&pull_data->refspec);
+  g_clear_pointer (&pull_data->loop, g_main_loop_unref);
+}
+
+static void
+eos_pull_data_finalize_impl (EosPullData *pull_data)
+{
+  g_clear_pointer (&pull_data->error, g_error_free);
+  g_clear_pointer (&pull_data->server_uri, g_free);
+  g_clear_pointer (&pull_data->commit_id, g_free);
+}
+
+EOS_DEFINE_REFCOUNTED (EOS_PULL_DATA,
+                       EosPullData,
+                       eos_pull_data,
+                       eos_pull_data_dispose_impl,
+                       eos_pull_data_finalize_impl)
+
+static EosPullData *
+eos_pull_data_new (GMainLoop *loop,
+                   const gchar *server_uri,
+                   EosRefspec *refspec,
+                   const gchar *commit_id,
+                   OstreeAsyncProgress *progress)
+{
+  EosPullData *pull_data = g_object_new (EOS_TYPE_PULL_DATA, NULL);
+
+  pull_data->loop = g_main_loop_ref (loop);
+  pull_data->server_uri = g_strdup (server_uri);
+  pull_data->refspec = g_object_ref (refspec);
+  pull_data->commit_id = g_strdup (commit_id);
+  pull_data->progress = g_object_ref (progress);
+
+  return pull_data;
+}
+
+static void
+pull_ready (GObject *object,
+            GAsyncResult *res,
+            gpointer pull_data_ptr)
+{
+  g_autoptr(EosPullData) pull_data = EOS_PULL_DATA (pull_data_ptr);
+
+  g_task_propagate_boolean (G_TASK (res), &pull_data->error);
+  g_main_loop_quit (pull_data->loop);
+}
+
+static void
+run_pull_task_func (GTask *task,
+                    gpointer repo_ptr,
+                    gpointer pull_data_ptr,
+                    GCancellable *cancellable)
+{
+  EosPullData *pull_data = EOS_PULL_DATA (pull_data_ptr);
+  OstreeRepo *repo = OSTREE_REPO (repo_ptr);
+  g_auto(GVariantBuilder) builder = { { { 0, } } };
+  g_autoptr(GVariant) options = NULL;
+  g_autoptr(GError) local_error = NULL;
+  const gchar *refs[] = { pull_data->refspec->ref };
+  const gchar *override_commit_ids[] = { pull_data->commit_id };
+  g_autoptr(ScopedMainContext) context = scoped_main_context_new ();
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE_VARDICT);
+  g_variant_builder_add (&builder,
+                         "{s@v}",
+                         "override-url",
+                         g_variant_new_variant (g_variant_new_string (pull_data->server_uri)));
+  g_variant_builder_add (&builder,
+                         "{s@v}",
+                         "refs",
+                         g_variant_new_variant (g_variant_new_strv (refs, 1)));
+  g_variant_builder_add (&builder,
+                         "{s@v}",
+                         "override-commit-ids",
+                         g_variant_new_variant (g_variant_new_strv (override_commit_ids, 1)));
+  g_variant_builder_add (&builder,
+                         "{s@v}",
+                         "depth",
+                         g_variant_new_variant (g_variant_new_int32 (0)));
```

Comment:
The mirroring mode does a couple things:

1. If not refs to pull are specified, then all refs on the remote are pulled. That's not (to my knowledge) the case here as we're specifying refs.

2. Refs are pulled into the local namespace (refs/heads) rather than the remote namespace (refs/remotes).

3. The summary file and signature are copied from upstream when they're available.

4. If you're mirroring into an archive-z2 repo, static deltas are skipped as you want to pull the exact objects rather than rebuilding them with deltas.